File: Algorithm/2023.10/Chulhyunjo/코딩테스트 기출문제/231019/셔틀버스.js
// 문자열 -> 숫자 (시간 -> 분)
function formatDate(time){
  time = time.split(':')
  let second = parseInt(time[0])*60 + parseInt(time[1])

  return second
}

// 숫자 -> 문자열 (분 -> 시간)
function formatDate2(time){
  let hour = Math.floor(time / 60)
  let minute = time % 60
  hour = hour < 10 ? '0' + hour : hour.toString()
  minute = minute < 10 ? '0'+minute : minute.toString()
  return hour + ':' + minute
}

function solution(n, t, m, timetable) {
  for(let i=0; i < timetable.length; i++){
      timetable[i] = formatDate(timetable[i])
  }
  timetable = timetable.filter((t)=> t!==1439)
  timetable.sort((a, b) => a - b)
  
  let firstTime = 540
  let cnt = 0
  let canUse = -1
  let timetable_cnt = timetable.length
  let used_cnt = 0
  
  while (cnt < n){
      let now_cnt = 0
  
      while (now_cnt < m && timetable_cnt){
          if(timetable[0] <= firstTime){
              canUse = timetable.shift()
              timetable_cnt--
          }
          now_cnt++
      }
      used_cnt += now_cnt
      cnt++
      firstTime += t
  }
  console.log(timetable_cnt, used_cnt, canUse)
  if (canUse === -1 || timetable_cnt === 0 && used_cnt < m * n){
      return formatDate2(540 + t * (n-1))
  }
  
  return formatDate2(canUse - 1)
}

File: Algorithm/2023.10/graywhales/셔틀버스.java
import java.util.*;
class Solution {
    static int n,t,m;
    static int[] times;
    public String solution(int n, int t, int m, String[] timetable) {
        
        this.n=n;
        this.t=t;
        this.m=m;
        
        int[] bus=new int[n*m];
        int[] visited=new int[n*m];
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                bus[i*m+j]=60*9+t*i;
            }
        }
        int p=timetable.length;
        times=new int[p];
        for(int i=0;i<p;i++){
            times[i]=Integer.parseInt(timetable[i].split(":")[0])*60+Integer.parseInt(timetable[i].split(":")[1]);
        }
        Arrays.sort(times);
        int i=0,j=0;
        while(i<n*m&&j<p){//버스 탑승가능한 사람 태우고 그사람의 시간을 visited에 기록
            if(bus[i]>=times[j]){
                visited[i++]=times[j++];
            }else{
                i++;
            }
        }
        int possible=0;
        int max=visited[n*m-1];
        
        if(visited[n*m-1]==0){
            return numToTime(bus[n*m-1]);//만약 마지막 버스 마지막자리 비었다면 그게 답.
        }else{
            for(i=n*m-1;i>=0;i--){
                if(visited[i]<max){//마지막 자리에 탄사람보다 시간이 적은사람이 있다면 그 다음자리에 탑승가능
                    possible=Math.min(bus[i+1],max-1);
                    break;
                }
            }
        }
        possible=Math.max(possible,max-1);
        
        return numToTime(possible);
    }
    public String numToTime(int n){
        int hour=n/60;
        int min=n%60;
        String time="";
        if(hour<10) time+="0"+hour;
        else time+=hour;
        time+=":";
        if(min<10) time+="0"+min;
        else time+=min;
        return time;
    }
}
